core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala

```diff
@@ -889,21 +891,40 @@ private class PythonAccumulatorParam(@transient private val serverHost: String,
     socket
   }
 
-  override def zero(value: JList[Array[Byte]]): JList[Array[Byte]] = new JArrayList
+  override def reset(): Unit = {
+    this._acc = new JArrayList[Array[Byte]]
+  }
+
+  override def isZero: Boolean = {
+    this._acc.isEmpty
+  }
 
-  override def addInPlace(val1: JList[Array[Byte]], val2: JList[Array[Byte]])
-      : JList[Array[Byte]] = synchronized {
+  override def copy(): PythonAccumulatorV2 = {
+    val newAcc = new PythonAccumulatorV2(serverHost, serverPort)
+    newAcc._acc = this._acc
+    newAcc
+  }
+
+  // This happens on the worker node, where we just want to remember all the updates
+  override def add(val2: JList[Array[Byte]]): Unit = synchronized {
```

Comment:
So this is the only access that happens on the worker side - all of the other access happen on the driver side where there is only one thread doing the merging.

Although if we started to do merging on the workers this could change so I'll make it so that the `merge` goes through  the `add` function if its ever triggered on a worker.